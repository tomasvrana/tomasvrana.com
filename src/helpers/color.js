import TinyColor from '@ctrl/tinycolor'
import logger from './logger'

const makeTransparent = (color, alpha) => {
  try {
    const c = new TinyColor(color)
    c.setAlpha(alpha)
    return c.toRgbString()
  } catch (error) {
    logger.error({
      description: 'error making color transparent',
      args: [ color, alpha ],
      error
    })
    return '#000000'
  }
}

export {
  makeTransparent
}
