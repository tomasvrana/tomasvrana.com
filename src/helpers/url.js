import { triggerEvent } from './dom'
import config from '../config'

const window = global.window || { location: { pathname: '/', href: '', origin: '', hash: '', hostname: '' }, history: {} }

export const getURIPath = () => window.location.pathname
export const getURIPathWithoutLang = lang => getURIPath().replace(new RegExp(`^\\/${lang}`), '')
export const languageSetInURL = (url = null) => !!config.availableLanguages[(url || getURIPath()).split('/').splice(1)[0]]
export const getHref = (path = null) => path ? `${window.location.origin}${path}` : window.location.href
export const setHref = href => (window.location.href = href)
export const isRelative = url => !url.match(/^(\/\/|https?:\/\/|mailto:)/)
export const isPath = url => !url.match(/^(\/|\/[^\\/].*)$/)
export const getHash = () => window.location.hash
export const getHostname = () => window.location.hostname

const getHashParams = () => {
  const currentHash = getHash().replace(/^#/, '')
  const params = currentHash.split('&').map(p => ({ [p.split('=')[0]]: p.split('=')[1] }))
  let paramsObj = {}
  params.forEach(p => (paramsObj = { ...paramsObj, ...p }))
  return paramsObj
}

export const setHashParam = (key, value) => {
  const params = getHashParams()
  if (!value) {
    delete params[key]
  } else {
    params[key] = value
  }

  const paramStrings = Object.keys(params).map(k => params[k] && k ? `${k}=${params[k]}` : '').filter(v => !!v)
  if (paramStrings.length > 0) {
    if (window.history.pushState) {
      window.history.pushState(null, null, `#${paramStrings.join('=')}`)
      triggerEvent({ type: 'hashchange' })
    } else {
      window.location.hash = `#${paramStrings.join('=')}`
    }
  } else {
    if (window.history.pushState) {
      window.history.pushState(null, null, '#')
      triggerEvent({ type: 'hashchange' })
    } else {
      window.location.hash = ''
    }
  }
}

export const getHashParam = key => {
  return getHashParams()[key] || ''
}
