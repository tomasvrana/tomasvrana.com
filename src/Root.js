import React, { Fragment, useEffect } from 'react'
import './bootstrap'
import PropTypes from 'prop-types'
import parseURL from 'url-parse'
import { getHref, setHref } from './helpers/url'
import { autoCapture } from './helpers/analytics'
import { REDIRECT } from './constants/analytics'

function adminRedirect () {
  const { protocol, slashes, host, hash } = parseURL(getHref())
  const token = (hash.replace(/^#\/?/, '').split('&').map(p => p.split('=')).filter(p => p[0] === 'invite_token').shift() || [])[1]
  if (token) {
    autoCapture({ category: REDIRECT, action: 'redirect_to_admin' })
    setHref(`${protocol}${slashes && '//'}${host}/admin/#/invite_token=${token}`)
  }
}

const AdminRedirector = () => {
  useEffect(() => {
    adminRedirect()
  }, [])

  return null
}

const Root = ({ element }) => (
  <Fragment>
    <AdminRedirector />
    {element}
  </Fragment>
)

Root.propTypes = {
  element: PropTypes.node.isRequired
}

export default Root
