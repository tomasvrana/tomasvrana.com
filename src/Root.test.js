import React from 'react'
import { mount } from 'enzyme'
import Root from './Root'
import * as url from './helpers/url'
import * as analytics from './helpers/analytics'

const getWrapper = props => {
  return mount(
    <Root
      element={(<p>Element</p>)}
      {...props}
    />
  )
}

describe('Root', () => {
  beforeEach(() => {
    analytics.autoCapture = jest.fn()
    url.getHref = jest.fn(() => 'https://example.org/')
    url.setHref = jest.fn()
  })

  test('it renders correctly', () => {
    expect(getWrapper()).toMatchSnapshot()
  })

  describe('when an invite token is not present in the URL', () => {
    beforeEach(() => {
      getWrapper()
    })

    test('it does not attempt to redirect to admin', () => {
      expect(url.setHref).not.toBeCalled()
    })
  })

  describe('when an invite token is present in the URL', () => {
    beforeEach(() => {
      url.getHref.mockReturnValue('https://example.org/#/invite_token=abcde')
    })

    test('it redirects to admin', () => {
      getWrapper()
      expect(url.setHref).toBeCalledWith('https://example.org/admin/#/invite_token=abcde')
    })
  })
})
