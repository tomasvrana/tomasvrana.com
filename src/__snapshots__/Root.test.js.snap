// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root it renders correctly 1`] = `
<Root
  element={
    <p>
      Element
    </p>
  }
>
  <AdminRedirector />
  <p>
    Element
  </p>
</Root>
`;
